tests/end-to-end/external-solution-integration/pkg/testsuite/deploy_function.go

```diff
@@ -0,0 +1,196 @@
+	
+	const gateway = getGateway()
+    if (gateway === undefined) {
+		throw new Error("gateway is undefined")	
+	}
+    if (event["data"] !== expectedPayload) {	
+		throw new Error("Payload not as expected")	
+    }
+	return new Promise((resolve, reject) => {	
+		const url = gateway + "/counter";	
+		console.log("Counter URL: ", url);	
+		sendReq(url, resolve, reject);	
+	}).then(resolved, rejected);	
+} };	
+`
+const functionDeps = `{"dependencies":{"request": "^2.88.0", "circular-json": "^0.5.9"}}`
+
+// DeployFunction deploys function to the cluster. The function will do PUT /counter to connected application upon receiving
+// an event
+type DeployFunction struct {
+	client          dynamic.ResourceInterface
+	name            string
+	port            int
+	expectedPayload string
+	legacy          bool
+}
+
+var _ step.Step = &DeployFunction{}
+
+func NewDeployFunction(name, expectedPayload string, port int, client dynamic.ResourceInterface, legacy bool) *DeployFunction {
+	return &DeployFunction{
+		client:          client,
+		name:            name,
+		port:            port,
+		legacy:          legacy,
+		expectedPayload: expectedPayload,
+	}
+}
+
+func (s *DeployFunction) Name() string {
+	return fmt.Sprintf("Deploy function %s", s.name)
+}
+
+func (s *DeployFunction) Run() error {
+	if err := s.createFunction(); err != nil {
+		return err
+	}
+
+	if err := retry.Do(s.isFunctionReady); err != nil {
+		return errors.Wrap(err, "function not ready")
+	}
+
+	return nil
+}
+
+func (s *DeployFunction) createFunction() error {
+	function := &serverless.Function{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Function",
+			APIVersion: serverless.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: s.name,
+		},
+		Spec: serverless.FunctionSpec{
+			Source: fmt.Sprintf(functionFmt, s.expectedPayload, s.legacy),
+			Deps:   functionDeps,
+		},
+	}
+
+	unstructuredObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(function)
+	if err != nil {
+		return err
+	}
+
+	_, err = s.client.Create(&unstructured.Unstructured{Object: unstructuredObj}, metav1.CreateOptions{})
+	return err
+}
+
+func (s *DeployFunction) Cleanup() error {
+	err := s.client.Delete(s.name, &metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	return helpers.AwaitResourceDeleted(func() (interface{}, error) {
+		return s.client.Get(s.name, metav1.GetOptions{})
+	})
+}
+
+func (s *DeployFunction) isFunctionReady() error {
+	function, err := s.getFunction()
+	if err != nil {
+		return err
+	}
+
+	if !s.isReady(function) {
+		return errors.New("function is not ready yet")
+	}
+
+	return nil
+}
+
+func (s *DeployFunction) getFunction() (serverless.Function, error) {
+	functionUnstructed, err := s.client.Get(s.name, metav1.GetOptions{})
+	if err != nil {
+		return serverless.Function{}, err
+	}
+
+	if len(functionUnstructed.Object) == 0 {
+		return serverless.Function{}, errors.New("no function found")
+	}
```

Comment:
I'm not sure how unstructured client works here, but the one we use from kubebuilder would just return NotFound error earlier, so this check would be useless

but you can leave it here just to be sure

Reply:
So I will leave it